redwoodjs/redwood
# Add deploy support for Heroku

## Add heroku deployment support for redwood apps.

---

### Core additions
- adds `rw deploy heroku` command
- creates heroku app with related accouterments
	- adds postgres instance [see notes](#notes)
	- adds node with nginx dyno (single dyno)
- adds scripts and config for heroku in `./config` and `./scripts`
- adds main heroku `Procfile` [see notes](#notes)
- adds PM2 and configuration
- updates `prisma.schema` to enable postgres [see notes](#notes)
- allows individual steps to be run
- slick TUI
- coverage north of 80%

---

### Additional changes
- "campground" refactors
	- start module pattern for deploy commands. [more info](https://en.wikipedia.org/wiki/Module_pattern)
	- refactor / add tests for adjacent / obviously duplicated code
	- typescript
- unrelated improvements / module upgrades / plugins
	- add watched builds for debugging specifically
	- separate some package scripts for better composability (none of the previous scripts functionally changed)
	- typescript exclusions for local dev
	- Add volta manager config to package.json (highly recommend) [more info](https://volta.sh/)


### More info can be found in the [readme](https://github.com/CodingZeal/redwood/blob/feat/cli/deploy/heroku/packages/cli/src/commands/deploy/modules/heroku/README.md) included in the heroku module

---

### Local Testing Instructions

- System requirements
	- POSIX complient shell (os x / linux)
	- heroku cli `brew tap heroku/brew && brew install heroku`
	- An Heroku account. (due to the pricing changes you may need a payment method on file)

---

### Steps
1. Pull / Clone latest redwoodjs code.
2. Switch to this branch `git checkout feat/cli/deploy/heroku`
3. Install latest deps `yarn install`
4. Build entire redwood project `yarn build`
5. Enter the `package/cli` module `cd package/cli`
6. Run the deploy command against a selected redwood project using the `RWJS_CWD` env var.
	1. `RWJS_CWD=../relative/path/to/redwood-project yarn dev deploy heroku`
	2. Run with `--help` to see avail options.
7. Watch the build process, then visit your site when complete
8. The entire heroku app and dependents can be destroyed automatically with the command: `yarn dev deploy heroku --destroy <your-app-name>`

---

### Notes:
- At the moment, only postgres is supported. I have another PR incoming with a complete docker setup that includes postgres setup for localdev and cloud services / frameworks
- The heroku `Procfile` is required for this deployment. This may be different than people are used to with node deploys that have used the automatic script execution. This is due to the nginx-buildpack which requires it and is the main entrypoint.
- I wanted to at least visibility on this for feedback and possibly added as an experimental feature before I bring in another PR as I am not familiar with your requirements as far as quality, patterns, and personalities are concerned.
- More detailed info about the service can be found in `packages/cli/src/commands/deploy/modules/heroku/README`

---

### TODO:
- Database instructions / options. (Maybe a separate PR) at least add documentation / manual instructions.
- Official documentation
- More fined grain control over created services with a config file. Unfortunately, the official app manifest requires a complete docker setup (in progress)
